packages/mural-client/src/fetch.ts

```diff
@@ -47,61 +48,70 @@ let fetchConfig: AuthenticatedFetchConfig;
 
 export const authenticated = () => {
   const session = fetchConfig.sessionStore.get();
-  return !!(session && !isTokenExpired(session.refreshToken));
+  return !!(session && !isTokenExpiredOrWillSoon(session.refreshToken));
 };
 
 const authenticatedFetch = async (
   input: RequestInfo,
   init: RequestInit = {},
 ): Promise<Response> => {
-  // setup all the handlers
-  await verifyTokensExpiration();
-
+  const session = await getSession();
   const initWithAuth = {
     ...init,
-    headers: withAuthenticationToken(init.headers || {}),
+    headers: {
+      ...init.headers,
+      ...(session ? { Authorization: `Bearer ${session?.accessToken}` } : {}),
+    },
   };
 
   try {
     const response = await fetch(input, initWithAuth);
     return await checkStatus(response);
   } catch (err: any) {
-    const errorHandler = catchAuthenticationError(input, initWithAuth);
+    const errorHandler = catchAuthenticationError();
     return errorHandler(err);
   }
 };
 
-function withAuthenticationToken(headers: HeadersInit) {
+const getSession = (): Promise<Session | null> => {
   const session = fetchConfig.sessionStore.get();
-  const accessToken = session && session.accessToken;
 
-  if (!accessToken) return headers;
-  return {
-    ...headers,
-    Authorization: `Bearer ${accessToken}`,
-  };
-}
+  if (!session) return Promise.resolve(null);
 
-async function verifyTokensExpiration() {
-  const session = fetchConfig.sessionStore.get();
-  if (session) {
-    if (isTokenExpired(session.refreshToken)) {
-      fetchConfig.sessionStore.delete();
-    } else if (isTokenExpired(session.accessToken)) {
-      await fetchConfig.refreshTokenFn({ store: true });
-    }
+  if (isTokenExpiredOrWillSoon(session.refreshToken)) {
+    fetchConfig.sessionStore.delete();
+    return Promise.resolve(null);
   }
-}
 
-function isTokenExpired(token: string) {
+  return handleTokenRefresh(session);
+};
+
+const handleTokenRefresh = memoize(
+  (session: Session) => {
+    if (isTokenExpiredOrWillSoon(session.accessToken)) {
+      return fetchConfig.refreshTokenFn({ store: true });
+    }
+    return Promise.resolve(session);
+  },
+  (session: Session) => session.accessToken,
+);
+
+const MINIMUM_TIME_FOR_REQUEST = 5000; // In milliseconds
```

Comment:
**structure**: move at the top

Reply:
**concern**: I think this isn't aggressive enough with the state of our API performance.

Our tokens currently have a 15m shelf-life, I think we can safely assume that the shelf-life of any token should be much larger than one individual request time.

With that said, we could use a fraction of that, say 5%, which should be PLENTY of room for any long running request. In that case 15m*60s/m * 0.05 = 45s = 45000ms

Reply:
Allowing 5 seconds should provide sufficient time to ensure that the next request "token validation" will be successful